File: python/open_guardrail/guards/regex_bomb.py
"""Detect ReDoS (Regular Expression Denial of Service) patterns."""
from __future__ import annotations

import re
import time
from open_guardrail.core import GuardResult

_PATTERNS: list[re.Pattern[str]] = [
    re.compile(r"\([^)]*[+*]\)[+*]"),
    re.compile(r"\([^)]*\|[^)]*\)\{"),
    re.compile(r"\(\.\*\)\{"),
    re.compile(r"\([^)]*[+*]\)\{[0-9,]+\}"),
    re.compile(r"\(\?:[^)]*[+*]\)[+*]"),
    re.compile(r"\([^)]*\+\)\+"),
    re.compile(r"\([^)]*\*\)\*"),
    re.compile(r"\(\.\+\)\+"),
    re.compile(r"\(\.\*\)\+"),
]


class _RegexBomb:
    def __init__(self, *, action: str = "block") -> None:
        self.name = "regex-bomb"
        self.action = action
        self._patterns = list(_PATTERNS)

    def check(self, text: str, stage: str = "input") -> GuardResult:
        start = time.perf_counter()
        matched: list[str] = []
        for pat in self._patterns:
            if pat.search(text):
                matched.append(pat.pattern)
        triggered = len(matched) > 0
        score = min(len(matched) / 3, 1.0) if triggered else 0.0
        elapsed = (time.perf_counter() - start) * 1000
        return GuardResult(
            guard_name="regex-bomb",
            passed=not triggered,
            action=self.action if triggered else "allow",
            score=score,
            message="ReDoS pattern detected" if triggered else None,
            latency_ms=round(elapsed, 2),
            details=(
                {"matched_patterns": len(matched)}
                if triggered
                else None
            ),
        )


def regex_bomb(*, action: str = "block") -> _RegexBomb:
    return _RegexBomb(action=action)

File: packages/guards/src/regex-bomb.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface RegexBombOptions {
  action: 'block' | 'warn';
}

const PATTERNS: RegExp[] = [
  /\([^)]*[+*]\)[+*]/,
  /\([^)]*\|[^)]*\)\{/,
  /\(\.\*\)\{/,
  /\([^)]*[+*]\)\{[0-9,]+\}/,
  /\(\?:[^)]*[+*]\)[+*]/,
  /\([^)]*\+\)\+/,
  /\([^)]*\*\)\*/,
  /\(\.\+\)\+/,
  /\(\.\*\)\+/,
];

export function regexBomb(options: RegexBombOptions): Guard {
  return {
    name: 'regex-bomb',
    version: '0.1.0',
    description: 'Detects ReDoS patterns that cause catastrophic backtracking',
    category: 'security',
    supportedStages: ['input'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const matched: string[] = [];

      for (const pattern of PATTERNS) {
        if (pattern.test(text)) {
          matched.push(pattern.source);
        }
      }

      const triggered = matched.length > 0;
      const score = triggered ? Math.min(matched.length / 3, 1.0) : 0;

      return {
        guardName: 'regex-bomb',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        score,
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { matchedPatterns: matched.length } : undefined,
      };
    },
  };
}
